chat-plugins/mafia.js

```diff
@@ -495,16 +501,36 @@ class MafiaTracker extends Rooms.RoomGame {
 		return true;
 	}
 
-	revive(deadPlayer) {
-		if (!this.started) return false;
-		if (!(deadPlayer in this.dead)) return false;
-		deadPlayer = this.dead[deadPlayer];
-		if (deadPlayer.treestump) delete deadPlayer.treestump;
-		if (deadPlayer.restless) delete deadPlayer.restless;
-		this.sendRoom(`${deadPlayer.name} was revived!`, {declare: true});
-		this.players[deadPlayer.userid] = deadPlayer;
-		this.roles.push(deadPlayer.role);
-		delete this.dead[deadPlayer.userid];
+	revive(user, deadPlayer) {
+		if (deadPlayer in this.players) return user.sendTo(this.room, `|error|The user ${deadPlayer} is already a living player.`);
+		if (deadPlayer in this.dead) {
+			deadPlayer = this.dead[deadPlayer];
+			if (deadPlayer.treestump) delete deadPlayer.treestump;
+			if (deadPlayer.restless) delete deadPlayer.restless;
+			this.sendRoom(`${deadPlayer.name} was revived!`, {declare: true});
+			this.players[deadPlayer.userid] = deadPlayer;
+			this.roles.push(deadPlayer.role);
+			delete this.dead[deadPlayer.userid];
+		} else {
+			const targetUser = Users(deadPlayer);
+			if (!targetUser || !targetUser.connected) return user.sendTo(this.room, `|error|The user "${deadPlayer}" was not found.`);
+			if (!this.room.users[targetUser.userid]) return this.errorReply(`${targetUser.name} is not in this room, and cannot be hosted.`);
+			let player = this.makePlayer(targetUser);
+			if (this.started) {
+				player.lynching = '';
+				player.role = {
+					name: `No set role`,
+					id: `nosetrole`,
+					alignment: 'solo',
+					memo: [`You were added to the game after it had started. To learn about your role, PM the host (${this.host}).`],
+				};
+				this.roles.push(player.role);
```

Comment:
update the role string too

Reply:
that part doesnt return, the role string, playercount etc are done at 533